src/number_theory/bernoulli.lean

```diff
@@ -1,81 +1,148 @@
 /-
 Copyright (c) 2020 Johan Commelin. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Johan Commelin
+Authors: Johan Commelin, Kevin Buzzard
 -/
 import data.rat
 import data.fintype.card
 
 /-!
 # Bernoulli numbers
 
-The Bernoulli numbers are a sequence of numbers that frequently show up in number theory.
+The Bernoulli numbers are a sequence of rational numbers that frequently show up in
+number theory.
 
-For example, they show up in the Taylor series of many trigonometric and hyperbolic functions,
-and also as (integral multiples of products of powers of `π` and)
-special values of the Riemann zeta function.
-(Note: these facts are not yet available in mathlib)
+## Mathematical overview
 
-In this file, we provide the definition,
-and the basic fact (`sum_bernoulli`) that
-$$ \sum_{k < n} \binom{n}{k} * B_k = n, $$
-where $B_k$ denotes the the $k$-th Bernoulli number.
+The Bernoulli numbers $(B_0, B_1, B_2, \ldots)=(1, 1/2, 1/6, 0, -1/30, \ldots)$ are
+a sequence of rational numbers. They show up in the formula for the sums of $k$th
+powers. They are related to the Taylor series expansions of $x/\tan(x)$ and
+of $\coth(x)$, and also show up in the values that the Riemann Zeta function
+takes both at both negative and positive integers (and hence in the
+theory of modular forms). For example If $1 \leq n$ is even then
 
+$$\zeta(2n)=\sum_{t\geq1}t^{-2n}=(-1)^{n+1}\frac{(2\pi)^{2n}B_{2n}}{2(2n)!}.$$
+
+Note however that this result is not yet formalised in Lean.
+
+The Bernoulli numbers can be formally defined using the power series
+
+$$\sum B_n\frac{t^n}{n!}=\frac{t}{1-e^{-t}}$$
+
+although that happens to not be the definition in mathlib (this is an *implementation
+detail* though, and need not concern the mathematician).
+
+Note that $B_1=+1/2$, meaning that we are using the $B_n^+$ of
+https://en.wikipedia.org/wiki/Bernoulli_number .
+To get the "minus" convention, just use `(-1)^n * bernoulli n`.
+
+There is no particular reason that the `+` convention was used.
+In some sense it's like choosing whether you want to sum over `fin n`
+(so `j < n`) or sum over `j ≤ n` (or `nat.antidiagonal n`). Indeed
+$$(t+1)\sum_{j\lt n}j^t=\sum_{k\leq t}\binom{t+1}{k}B_k^{-}n^{t+1-k}$$
+and
+$$(t+1)\sum_{j\leq n}j^t=\sum_{k\leq t}\binom{t+1}{k}B_k^{+}n^{t+1-k}.$$
+
+## Implementation detail
+
+```
+def bernoulli : ℕ → ℚ :=
+well_founded.fix nat.lt_wf
+  (λ n bernoulli, 1 - ∑ k : fin n, (n.choose k) * bernoulli k k.2 / (n + 1 - k))
+```
+
+Setting `n = 0` we deduce that `bernoulli 0 = 1` and one can go on from there.
```

Comment:
Is there any value in duplicating the source of the definition here? Maybe this last line should just go in the docstring for `bernoulli` below.

Reply:
I've made this part of the module docstring better.